File: packages/core/src/keys.ts
/**
 * Ed25519 key helpers for Arky. The signing key is a 32-byte seed; the matching
 * verification identity is a `did:key:z6Mk…` derived from the public key so that
 * `identity.id` resolves to the key that signed the TIM (ARKY-TIM-v1 §6.1).
 *
 * Using these together guarantees the DID and the key agree — the common
 * footgun when hand-constructing identities.
 */

import { ed25519 } from '@noble/curves/ed25519';
import { base58btcEncode } from './cid.ts';

export interface KeyPair {
  /** 32-byte Ed25519 seed (private). Keep secret. */
  privateKey: Uint8Array;
  /** 32-byte Ed25519 public key. */
  publicKey: Uint8Array;
  /** did:key identity that resolves to `publicKey` (use as identity.id). */
  did: string;
}

/** Derive the did:key (multicodec 0xed01 + pubkey, base58btc, 'z' prefix). */
export function didKeyFromPublicKey(publicKey: Uint8Array): string {
  const mc = new Uint8Array(2 + publicKey.length);
  mc[0] = 0xed;
  mc[1] = 0x01;
  mc.set(publicKey, 2);
  return 'did:key:z' + base58btcEncode(mc);
}

/** Public key + did:key for a given 32-byte seed. */
export function fromSeed(privateKey: Uint8Array): KeyPair {
  const publicKey = ed25519.getPublicKey(privateKey);
  return { privateKey, publicKey, did: didKeyFromPublicKey(publicKey) };
}

/** Generate a fresh random Ed25519 keypair with its matching did:key. */
export function generateKeyPair(): KeyPair {
  return fromSeed(ed25519.utils.randomPrivateKey());
}

File: packages/core-rs/src/keys.rs
//! Ed25519 key helpers. The matching verification identity is a `did:key:z6Mk…`
//! derived from the public key, so `identity.id` resolves to the signing key
//! (ARKY-TIM-v1 §6.1). Using these together guarantees the DID and key agree.

use crate::cid::to_multibase;
use ed25519_dalek::SigningKey;

/// An Ed25519 keypair plus its matching did:key identity.
pub struct KeyPair {
    pub signing_key: SigningKey,
    pub did: String,
}

/// Derive the did:key (multicodec 0xed01 + pubkey, base58btc, 'z' prefix).
pub fn did_key_from_public(public_key: &[u8]) -> String {
    let mut mc = Vec::with_capacity(2 + public_key.len());
    mc.push(0xed);
    mc.push(0x01);
    mc.extend_from_slice(public_key);
    // to_multibase yields "z<base58>"; the did:key form is "did:key:z<base58>".
    format!("did:key:{}", to_multibase(&mc))
}

/// KeyPair from a 32-byte seed.
pub fn from_seed(seed: [u8; 32]) -> KeyPair {
    let signing_key = SigningKey::from_bytes(&seed);
    let did = did_key_from_public(signing_key.verifying_key().as_bytes());
    KeyPair { signing_key, did }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn did_matches_key() {
        // The did:key derived from a known seed's public key starts with z6Mk
        // and round-trips back to the same 32-byte key via resolve_did_key.
        let kp = from_seed([7u8; 32]);
        assert!(kp.did.starts_with("did:key:z6Mk"));
        let resolved = crate::tim::resolve_did_key(&kp.did).unwrap();
        assert_eq!(resolved, kp.signing_key.verifying_key().as_bytes());
    }
}
